Index: amneziawg-linux-kernel-module-1.0.20260210/src/crypto/zinc/blake2s/blake2s.c
===================================================================
--- amneziawg-linux-kernel-module-1.0.20260210.orig/src/crypto/zinc/blake2s/blake2s.c
+++ amneziawg-linux-kernel-module-1.0.20260210/src/crypto/zinc/blake2s/blake2s.c
@@ -251,7 +251,7 @@ static int __init mod_init(void)
 {
 	if (!nosimd)
 		blake2s_fpu_init();
-	if (!selftest_run("blake2s", blake2s_selftest, blake2s_nobs,
+	if (!selftest_run("blake2s", blake2s_selftest, ARRAY_SIZE(blake2s_nobs) > 0 ? blake2s_nobs : NULL,
 			  ARRAY_SIZE(blake2s_nobs)))
 		return -ENOTRECOVERABLE;
 	return 0;
Index: amneziawg-linux-kernel-module-1.0.20260210/src/crypto/zinc/curve25519/curve25519.c
===================================================================
--- amneziawg-linux-kernel-module-1.0.20260210.orig/src/crypto/zinc/curve25519/curve25519.c
+++ amneziawg-linux-kernel-module-1.0.20260210/src/crypto/zinc/curve25519/curve25519.c
@@ -89,7 +89,7 @@ static int __init mod_init(void)
 {
 	if (!nosimd)
 		curve25519_fpu_init();
-	if (!selftest_run("curve25519", curve25519_selftest, curve25519_nobs,
+	if (!selftest_run("curve25519", curve25519_selftest, ARRAY_SIZE(curve25519_nobs) > 0 ? curve25519_nobs : NULL,
 			  ARRAY_SIZE(curve25519_nobs)))
 		return -ENOTRECOVERABLE;
 	return 0;
