Index: amneziawg-linux-kernel-module-amneziawg-ac946a9/src/crypto/zinc/blake2s/blake2s.c
===================================================================
--- amneziawg-linux-kernel-module-amneziawg-ac946a9.orig/src/crypto/zinc/blake2s/blake2s.c
+++ amneziawg-linux-kernel-module-amneziawg-ac946a9/src/crypto/zinc/blake2s/blake2s.c
@@ -251,7 +251,7 @@ static int __init mod_init(void)
 {
 	if (!nosimd)
 		blake2s_fpu_init();
-	if (!selftest_run("blake2s", blake2s_selftest, blake2s_nobs,
+	if (!selftest_run("blake2s", blake2s_selftest, ARRAY_SIZE(blake2s_nobs) > 0 ? blake2s_nobs : NULL,
 			  ARRAY_SIZE(blake2s_nobs)))
 		return -ENOTRECOVERABLE;
 	return 0;
Index: amneziawg-linux-kernel-module-amneziawg-ac946a9/src/crypto/zinc/curve25519/curve25519.c
===================================================================
--- amneziawg-linux-kernel-module-amneziawg-ac946a9.orig/src/crypto/zinc/curve25519/curve25519.c
+++ amneziawg-linux-kernel-module-amneziawg-ac946a9/src/crypto/zinc/curve25519/curve25519.c
@@ -89,7 +89,7 @@ static int __init mod_init(void)
 {
 	if (!nosimd)
 		curve25519_fpu_init();
-	if (!selftest_run("curve25519", curve25519_selftest, curve25519_nobs,
+	if (!selftest_run("curve25519", curve25519_selftest, ARRAY_SIZE(curve25519_nobs) > 0 ? curve25519_nobs : NULL,
 			  ARRAY_SIZE(curve25519_nobs)))
 		return -ENOTRECOVERABLE;
 	return 0;
Index: amneziawg-linux-kernel-module-amneziawg-ac946a9/src/crypto/zinc/selftest/blake2s.c
===================================================================
--- amneziawg-linux-kernel-module-amneziawg-ac946a9.orig/src/crypto/zinc/selftest/blake2s.c
+++ amneziawg-linux-kernel-module-amneziawg-ac946a9/src/crypto/zinc/selftest/blake2s.c
@@ -2072,7 +2072,11 @@ static bool __init blake2s_selftest(void
 		buf[i] = (u8)i;
 
 	for (i = 0; i < ARRAY_SIZE(blake2s_keyed_testvecs); ++i) {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 19, 0)
+		blake2s(key, BLAKE2S_HASH_SIZE, buf, i, hash, BLAKE2S_KEY_SIZE);
+#else
 		blake2s(hash, buf, key, BLAKE2S_HASH_SIZE, i, BLAKE2S_KEY_SIZE);
+#endif
 		if (memcmp(hash, blake2s_keyed_testvecs[i], BLAKE2S_HASH_SIZE)) {
 			pr_err("blake2s keyed self-test %zu: FAIL\n", i + 1);
 			success = false;
@@ -2080,7 +2084,11 @@ static bool __init blake2s_selftest(void
 	}
 
 	for (i = 0; i < ARRAY_SIZE(blake2s_testvecs); ++i) {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 19, 0)
+		blake2s(NULL, 0, buf, i, hash, BLAKE2S_HASH_SIZE);
+#else
 		blake2s(hash, buf, NULL, BLAKE2S_HASH_SIZE, i, 0);
+#endif
 		if (memcmp(hash, blake2s_testvecs[i], BLAKE2S_HASH_SIZE)) {
 			pr_err("blake2s unkeyed self-test %zu: FAIL\n", i + i);
 			success = false;
Index: amneziawg-linux-kernel-module-amneziawg-ac946a9/src/receive.c
===================================================================
--- amneziawg-linux-kernel-module-amneziawg-ac946a9.orig/src/receive.c
+++ amneziawg-linux-kernel-module-amneziawg-ac946a9/src/receive.c
@@ -535,10 +535,13 @@ void wg_packet_decrypt_worker(struct wor
 				PACKET_STATE_CRYPTED : PACKET_STATE_DEAD;
 		wg_queue_enqueue_per_peer_rx(skb, state);
 #ifdef COMPAT_CRYPTO_IS_ZINC
-		simd_relax(&simd_context);
-#endif
+		if( simd_relax(&simd_context) ||
+				(!(simd_context & HAVE_SIMD_IN_USE) && need_resched()))
+			cond_resched();
+#else
 		if (need_resched())
 			cond_resched();
+#endif
 	}
 
 #ifdef COMPAT_CRYPTO_IS_ZINC
Index: amneziawg-linux-kernel-module-amneziawg-ac946a9/src/send.c
===================================================================
--- amneziawg-linux-kernel-module-amneziawg-ac946a9.orig/src/send.c
+++ amneziawg-linux-kernel-module-amneziawg-ac946a9/src/send.c
@@ -368,10 +368,13 @@ void wg_packet_encrypt_worker(struct wor
 		wg_queue_enqueue_per_peer_tx(first, state);
 
 #ifdef COMPAT_CRYPTO_IS_ZINC
-		simd_relax(&simd_context);
-#endif
+		if( simd_relax(&simd_context) ||
+				(!(simd_context & HAVE_SIMD_IN_USE) && need_resched()))
+			cond_resched();
+#else
 		if (need_resched())
 			cond_resched();
+#endif
 	}
 #ifdef COMPAT_CRYPTO_IS_ZINC
 	simd_put(&simd_context);
