Index: ovpn-backports-ovpn-net-next-main-7.1.0-rc1-2026050800/compat-include/linux/overflow.h
===================================================================
--- ovpn-backports-ovpn-net-next-main-7.1.0-rc1-2026050800.orig/compat-include/linux/overflow.h
+++ ovpn-backports-ovpn-net-next-main-7.1.0-rc1-2026050800/compat-include/linux/overflow.h
@@ -4,7 +4,7 @@
 #include <linux/version.h>
 #include_next <linux/overflow.h>
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 321)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 14, 0)
 
 static inline size_t __must_check ovpn_size_mul(size_t factor1, size_t factor2)
 {
@@ -18,6 +18,6 @@ static inline size_t __must_check ovpn_s
 
 #define size_mul ovpn_size_mul
 
-#endif /* LINUX_VERSION_CODE < KERNEL_VERSION(4, 19, 321) */
+#endif /* LINUX_VERSION_CODE < KERNEL_VERSION(5, 14, 0) */
 
 #endif /* _NET_OVPN_LINUX_OVERFLOW_H_ */
